numbat/src/list.rs

```diff
@@ -0,0 +1,297 @@
+
+use std::{collections::VecDeque, fmt, sync::Arc};
+
+use crate::{value::Value, RuntimeError};
+
+/// Reference counted list / list view
+#[derive(Clone, Eq)]
+pub struct NumbatList<T> {
+    /// The original alloc shared between all values
+    alloc: Arc<VecDeque<T>>,
+    /// The indexes accessible to us. If `None` we own the whole allocation
+    view: Option<(usize, usize)>,
+}
+
+impl<T> Default for NumbatList<T> {
+    fn default() -> Self {
+        Self {
+            alloc: Default::default(),
+            view: Default::default(),
+        }
+    }
+}
+
+impl<T: fmt::Debug + Clone> fmt::Debug for NumbatList<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+impl<T: PartialEq> PartialEq for NumbatList<T> {
+    fn eq(&self, other: &Self) -> bool {
+        // Best case scenario, the slice don't have the same len and we can early exit
+        if self.len() != other.len() {
+            return false;
+        }
+        // Second best case scenario, the slice comes from the same allocation
+        // and have the same view => they are equal
+        if Arc::as_ptr(&self.alloc) == Arc::as_ptr(&other.alloc) && self.view == other.view {
+            true
+        } else {
+            // Worst case scenario, we need to compare all the elements one by one
+            self.iter().zip(other.iter()).all(|(l, r)| l == r)
+        }
+    }
+}
+
+impl<T> NumbatList<T> {
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        if let Some(view) = self.view {
+            view.1 - view.0
+        } else {
+            self.alloc.len()
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            alloc: Arc::new(VecDeque::with_capacity(capacity)),
+            view: None,
+        }
+    }
+
+    pub fn front(&self) -> Option<&T> {
+        if let Some(view) = self.view {
+            self.alloc.get(view.0)
+        } else {
+            self.alloc.front()
+        }
+    }
+
+    /// Advance the view you have of the list by one.
+    pub fn advance_view(&mut self) -> Result<(), RuntimeError> {
+        if self.is_empty() {
+            return Err(RuntimeError::EmptyList);
+        }
+        if let Some(view) = &mut self.view {
+            view.0 += 1;
+            // should be ensured by the if above
+            debug_assert!(view.0 <= view.1);
+        } else {
+            self.view = Some((1, self.len()));
+        }
+        Ok(())
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        self.alloc
+            .iter()
+            .skip(self.view.map_or(0, |(start, _end)| start))
+            .take(
+                self.view
+                    .map_or(self.alloc.len(), |(start, end)| end - start),
+            )
```

Comment:
I wonder if this would be a bit easier to read like this
```suggestion
        if let Some((start, end)) = self.view {
          self.alloc.iter().skip(start).take(end - start)
        } else {
          self.alloc.iter()
        }
```

or maybe even
```suggestion
        if let Some((start, end)) = self.view {
          self.alloc[start..end].iter()
        } else {
          self.alloc.iter()
        }
```
if that works?

Reply:
The first proposition doesn't work because the two iterators don't share the same type.
To make it work, I would need to wrap them in a `Box<dyn Iterator<...>>`, which is not ideal.
I’ll try the second one, but I think that's what I initially tried, and it wasn't working as well.
Will keep you updated

Reply:
Oh yes I remember now, the second solution doesn't work either because `VecDeque` doesn't implements `Index<SliceIndex>` like the `Vec` do, but only [`Index<usize>`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html#impl-Index%3Cusize%3E-for-VecDeque%3CT,+A%3E).

Which means you can only retrieve one element at a time in a `VecDeque` 😔 

I still tried to simplify it a little bit: https://github.com/sharkdp/numbat/pull/515/commits/56758d819870bb08c99929a3bbf3f178a7a57721